esmvalcore/_recipe/from_datasets.py

```diff
@@ -0,0 +1,327 @@
+"""Functions for creating/updating a recipe with `Dataset`s."""
+from __future__ import annotations
+
+import itertools
+import logging
+import re
+from copy import deepcopy
+from functools import partial
+from typing import TYPE_CHECKING, Any, Iterable, Mapping, Sequence
+
+from nested_lookup import nested_delete
+
+from esmvalcore.exceptions import RecipeError
+
+if TYPE_CHECKING:
+    from esmvalcore.dataset import Dataset
+
+logger = logging.getLogger(__name__)
+
+Recipe = dict[str, Any]
+Facets = dict[str, Any]
+
+
+def _datasets_to_raw_recipe(datasets: Iterable[Dataset]) -> Recipe:
+    """Convert datasets to a recipe dict."""
+    diagnostics: dict[str, dict[str, Any]] = {}
+
+    for dataset in datasets:
+        diagnostic_name: str = dataset.facets['diagnostic']  # type: ignore
+        if diagnostic_name not in diagnostics:
+            diagnostics[diagnostic_name] = {'variables': {}}
+        variables = diagnostics[diagnostic_name]['variables']
+        if 'variable_group' in dataset.facets:
+            variable_group = dataset.facets['variable_group']
+        else:
+            variable_group = dataset.facets['short_name']
+        if variable_group not in variables:
+            variables[variable_group] = {'additional_datasets': []}
+        facets: dict[str, Any] = dataset.minimal_facets
+        facets.pop('diagnostic', None)
+        if facets['short_name'] == variable_group:
+            facets.pop('short_name')
+        if dataset.ancillaries:
+            facets['ancillary_variables'] = []
+        for ancillary in dataset.ancillaries:
+            anc_facets = {}
+            for key, value in ancillary.minimal_facets.items():
+                if facets.get(key) != value:
+                    anc_facets[key] = value
+            facets['ancillary_variables'].append(anc_facets)
+        variables[variable_group]['additional_datasets'].append(facets)
+
+    recipe = {'diagnostics': diagnostics}
+    return recipe
+
+
+def _datasets_to_recipe(datasets: Iterable[Dataset]) -> Recipe:
+    """Convert datasets to a condensed recipe dict."""
+    for dataset in datasets:
+        if 'diagnostic' not in dataset.facets:
+            raise RecipeError(f"'diagnostic' facet missing from {dataset},"
```

Comment:
The dataset string description starts with the word `Dataset`, that's why I did not add it here because then the text will read `'diagnostic' facet missing from dataset Dataset: ..,`

Reply:
ah gotcha!